### PR TITLE
Add an example of object segmentation (ClipSeg)

### PR DESCRIPTION
**Pull request by cfgfung**
# What does this PR do?

The ClipSeg model has been validated using graph mode on Gaudi 2. Adding this as an example for users.

Performance number:

Without graph mode
  - n_iterations: 20
  - Total latency (ms): 670.0191497802734
  - Average latency (ms): **33.50095748901367**

With graph mode
  - n_iterations: 20
  - Total latency (ms): 95.40128707885742
  - Average latency (ms): **4.770064353942871**

Fixes # (issue)
N.A.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by cfgfung** [comment]
@regisss 
Please take a look at this one. This is a validated model and this is an example script for users.

**Comment by yeonsily** [comment]
@cfgfung I see that you have a couple of more PRs which has the same change.
https://github.com/huggingface/optimum-habana/pull/814/
https://github.com/huggingface/optimum-habana/pull/826

Which is the valid one? and can you please close any not valid PR?

**Comment by cfgfung** [comment]
> @cfgfung I see that you have a couple of more PRs which has the same change. #814 #826
> 
> Which is the valid one? and can you please close any not valid PR?

@yeonsily 
They are not the same. Each PR is for a dedicated model.  Each model requires a different output and dedicated preprocessing /postprocessing codes. They only share a similar start script.

**Comment by jiminha** [comment]
@cfgfung Your result looks good. Could you also port relevant test case for your model and make sure those are passing? 
https://github.com/huggingface/transformers/blob/main/tests/models/clipseg to https://github.com/huggingface/optimum-habana/tree/main/tests/transformers/tests/models

**Comment by jiminha** [comment]
> @cfgfung Your result looks good. Could you also port relevant test case for your model and make sure those are passing? https://github.com/huggingface/transformers/blob/main/tests/models/clipseg to https://github.com/huggingface/optimum-habana/tree/main/tests/transformers/tests/models

Actually, above model test cases are optional, it's good to check but not necessary. 
For the test cases for this ClipSeg model, please add similar test cases as your other PR(SAM).

**Comment by cfgfung** [comment]
> > @cfgfung Your result looks good. Could you also port relevant test case for your model and make sure those are passing? https://github.com/huggingface/transformers/blob/main/tests/models/clipseg to https://github.com/huggingface/optimum-habana/tree/main/tests/transformers/tests/models
> 
> Actually, above model test cases are optional, it's good to check but not necessary. For the test cases for this ClipSeg model, please add similar test cases as your other PR(SAM).

Hi, I have added the test and followed the same format as the PR of SAM.
Ran 'make style' and the unit test. Also aligned the folder structure and the readme.

**Comment by jiminha** [comment]
@cfgfung  all looks good. Could you please make sure that you ran "make style" and upload the fix? I still see many style error.

**Comment by cfgfung** [comment]
> @cfgfung all looks good. Could you please make sure that you ran "make style" and upload the fix? I still see many style error.

Hi,

There might be some updates in the rules and hence triggered the new errors. 
I have re-ran "make style" and the result is listed below.

image

**Review by jiminha — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -41,6 +41,11 @@ fast_tests_diffusers:
 	python -m pip install .[tests]
 	python -m pytest tests/test_diffusers.py
 
+# Run unit tests related to ClipSeg
+fast_tests_clipseg:
+	python -m pip install .[tests]
+	python -m pytest tests/test_clipseg.py
```
Comment (regisss) [inline Makefile:47]:
```suggestion
slow_tests_object_segmentation: test_installs
	python -m pytest tests/test_clipseg.py
```
Let's rather consider it as a slow test command as it contains a regression test and keep fast tests for generic stuff.

File: tests/test_clipseg.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        model, processor = self.prepare_model_and_processor()
+        texts, image = self.prepare_data()
+        inputs = processor(text=texts, images=[image] * len(texts), padding=True, return_tensors="pt").to("hpu")
+        outputs = model(**inputs)
+        probs = outputs.logits_per_image.softmax(dim=-1).detach().cpu().numpy()[0]
+        expected_scores = np.array([0.02889409, 0.87959206, 0.09151383])  # from CPU
+        self.assertEqual(len(probs), 3)
+        self.assertLess(np.abs(probs - expected_scores).max(), 0.01)
+
+    def test_inference_bf16(self):
```
Comment (regisss) [inline tests/test_clipseg.py:60]:
```suggestion
    def test_inference_autocast(self):
```
For inference, I prefer keeping the bf16 calling for cases where the whole model is in bf16.

File: tests/test_clipseg.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+        model = ht.hpu.wrap_in_hpu_graph(model)  # Apply graph
+
+        outputs = model(**inputs)
+        probs = outputs.logits_per_image.softmax(dim=-1).to(torch.float32).detach().cpu().numpy()[0]
+        expected_scores = np.array([0.02889409, 0.87959206, 0.09151383])  # from CPU
+        self.assertEqual(len(probs), 3)
+        self.assertEqual(probs.argmax(), expected_scores.argmax())
+
+    def test_no_latency_regression_bf16(self):
```
Comment (regisss) [inline tests/test_clipseg.py:85]:
```suggestion
    def test_no_latency_regression_autocast(self):
```

File: Makefile (hunk 1 above)
Comment (cfgfung) [reply Makefile:47]: Sure, that makes sense.

**Review by regisss — APPROVED** [review]

**Pull request merged**